.dotnet/src/Custom/Internal/ClientPipelineExtensions.cs

```diff
@@ -0,0 +1,85 @@
+using System;
+using System.ClientModel;
+using System.ClientModel.Primitives;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OpenAI;
+
+internal static partial class ClientPipelineExtensions
+{
+    // CUSTOM:
+    // - Supplemented exception body with deserialized OpenAI error details
+
+    public static async ValueTask<PipelineResponse> ProcessMessageAsync(
+        this ClientPipeline pipeline,
+        PipelineMessage message,
+        RequestOptions options)
+    {
+        await pipeline.SendAsync(message).ConfigureAwait(false);
+
+        if (message.Response == null)
+        {
+            throw new InvalidOperationException("Failed to receive Result.");
+        }
+
+        if (message.Response.IsError && (options?.ErrorOptions & ClientErrorBehaviors.NoThrow) != ClientErrorBehaviors.NoThrow)
+        {
+            throw await TryBufferResponseAndCreateErrorAsync(message.Response).ConfigureAwait(false) switch
+            {
+                string errorMessage when !string.IsNullOrEmpty(errorMessage)
+                    => new ClientResultException(errorMessage, message.Response),
+                _ => new ClientResultException(message.Response),
+            };
+        }
+
+        return message.Response;
+    }
+        
+    public static PipelineResponse ProcessMessage(
+        this ClientPipeline pipeline,
+        PipelineMessage message,
+        RequestOptions options)
+    {
+        pipeline.Send(message);
+
+        if (message.Response.IsError && (options?.ErrorOptions & ClientErrorBehaviors.NoThrow) != ClientErrorBehaviors.NoThrow)
+        {
+            throw TryBufferResponseAndCreateError(message.Response) switch
+            {
+                string errorMessage when !string.IsNullOrEmpty(errorMessage)
+                    => new ClientResultException(errorMessage, message.Response),
+                _ => new ClientResultException(message.Response),
+            };
+        }
+
+        return message.Response;
+    }
+
+    private static string TryBufferResponseAndCreateError(PipelineResponse response)
+    {
+        response.BufferContent();
```

Comment:
Would it make sense to wrap this in a check that message.BufferResponse is true?  If it's false, we could end up in a case where we try to buffer a very large network stream and throw out of memory exception.  If we can rule out this case -- i.e. that we would run out of memory -- it's probably fine.

Reply:
That's a great point. For at least this scenario-specific application, there's absolutely no reason I know of *not* to wrap it up as you suggest. I'll do that.